File: cpp/recursion/string_permutation.cpp
#include <iostream>
using namespace std;

void indent(int n){
  for(int i=0;i<n;i++){
    cout<<"    ";
  }
}

void permuteHelper(string soFar, string rest)
{
  indent(soFar.length());
  cout << "permute(\"" << soFar << "\",\""<< rest << "\")" << endl;
  if (rest.empty())
  {
    cout << soFar << endl;
  }
  else
  {
    for (int i = 0; i < rest.length(); i++)
    {

      //cout<<rest.substr(0, i)<<" "<<rest.substr(i + 1)<<endl;
      string remaining = rest.substr(0, i) + rest.substr(i + 1);
      //cout<<soFar+rest[i]<<" "<<remaining<<endl;
      permuteHelper(soFar + rest[i], remaining);
    }
  }
}

void permute(string s)
{
  permuteHelper("", s);
}

int main()
{
  permute("abcd");
}

File: java/recursion/StringPermutation.java

public class StringPermutation{
  public static void main(String[] args){
    permute("abc");
  }

  static void permute(String s){
    permuteHelper("",s);
  }

  static void permuteHelper(String soFar,String rest){
    indent(soFar.length());
    System.out.println("permuteHelper(\""+ soFar + "\",\""+rest+"\")");

    if(rest.length()==0){
      System.out.println(soFar);
    }else{
      for(int i=0;i<rest.length();i++){
        String temp=rest.substring(0,i) + rest.substring(i+1);
        permuteHelper(soFar + rest.charAt(i), temp);
      }
    }

  }

  static void indent(int n){
    for(int i=0;i<n;i++){
      System.out.print("    ");
    }
  }

}